backend/src/main/java/vn/edu/hust/nmcnpm_20242_n3/controller/FineController.java

```diff
@@ -0,0 +1,117 @@
+        }
+        if (fineDTO.getAmount() <= 0) {
+            throw new IllegalArgumentException("Fine amount must be greater than zero.");
+        }
+        if (fineDTO.getDescription() == null || fineDTO.getDescription().isEmpty()) {
+            throw new IllegalArgumentException("Description cannot be null or empty.");
+        }
+        User user = userRepository.findById(fineDTO.getUserId())
+                .orElseThrow(() -> new IllegalArgumentException("User not found"));
+        BookLoan bookLoan = bookLoanRepository.findById(fineDTO.getBookLoanId())
+                .orElseThrow(() -> new IllegalArgumentException("BookLoan not found"));
+        Fine fine = new Fine();
+        fine.setAmount(fineDTO.getAmount());
+        fine.setDescription(fineDTO.getDescription());
+        fine.setUser(user);
+        fine.setBookLoan(bookLoan);
+        Fine savedFine = fineService.addFine(fine);
+        return new ResponseEntity<>(FineDTO.fromEntity(savedFine), HttpStatus.CREATED);
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<FineDTO> updateFine(@PathVariable String id, @RequestBody FineDTO fineDTO) {
+        User user = userRepository.findById(fineDTO.getUserId())
+                .orElseThrow(() -> new IllegalArgumentException("User not found"));
+        BookLoan bookLoan = bookLoanRepository.findById(fineDTO.getBookLoanId())
+                .orElseThrow(() -> new IllegalArgumentException("BookLoan not found"));
+        Fine fine = new Fine();
+        fine.setId(id);
+        fine.setAmount(fineDTO.getAmount());
+        fine.setDescription(fineDTO.getDescription());
+        fine.setUser(user);
+        fine.setBookLoan(bookLoan);
+        Fine updatedFine = fineService.updateFine(id, fine);
+        return new ResponseEntity<>(FineDTO.fromEntity(updatedFine), HttpStatus.OK);
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<String> deleteFine(@PathVariable(name = "id") String id) {
+        fineService.deleteFine(id);
+        return new ResponseEntity<>("Fine deleted successfully", HttpStatus.OK);
+    }
+
+    @GetMapping("/user/{userId}")
+    public ResponseEntity<List<FineDTO>> getFinesByUserId(@PathVariable(name = "userId") String userId) {
+        List<FineDTO> dtos = fineService.getFinesByUserId(userId)
+                .stream()
+                .map(FineDTO::fromEntity)
+                .collect(Collectors.toList());
+        return new ResponseEntity<>(dtos, HttpStatus.OK);
+    }
+
+    @GetMapping("/date-range")
+    public ResponseEntity<List<FineDTO>> getFinesByDateRange(@RequestParam Date startDate, @RequestParam Date endDate) {
+        List<FineDTO> dtos = fineService.getFinesByDateRange(startDate, endDate)
+                .stream()
+                .map(FineDTO::fromEntity)
+                .collect(Collectors.toList());
+        return new ResponseEntity<>(dtos, HttpStatus.OK);
+    }
```

Comment:
This method wouldn't work as it will not recognize any date format. Should we exclusively give this a `@DateTimeFormat` for the params?